### PR TITLE
Fix the bug of AllReduceDepPass

### PR DESCRIPTION
**Pull request by chengduoZH**

### CODE REVIEW

**Review by gongweibao — DISMISSED** [review]
单测好像有问题呀。

File: paddle/fluid/framework/details/build_strategy.cc [hunk 1]
```diff
@@ -173,6 +169,9 @@ class ParallelExecutorPassBuilder : public ir::PassBuilder {
         VLOG(10) << "Add all_reduce_mode_multi_devices_pass";
         multi_devices_pass =
             AppendPass("all_reduce_mode_multi_devices_pass").get();
+        // experimental shows that the program will be faster if append
+        // all_reduce_deps_pass here.
```
Comment (panyx0718) [inline paddle/fluid/framework/details/build_strategy.cc]: how much faster?

File: paddle/fluid/framework/details/all_reduce_op_handle.cc [hunk 2]
```diff
@@ -21,7 +21,7 @@
 // asynchronous nccl allreduce or synchronous issue:
 // https://github.com/PaddlePaddle/Paddle/issues/15049
 DEFINE_bool(
-    sync_nccl_allreduce, false,
+    sync_nccl_allreduce, true,
```
Comment (panyx0718) [inline paddle/fluid/framework/details/all_reduce_op_handle.cc:31]: how much faster?

File: paddle/fluid/framework/details/all_reduce_deps_pass.cc [hunk 3]
```diff
@@ -23,34 +23,78 @@
...
+  for (size_t i = 1; i < all_reduce_op_handles.size(); ++i) {
+    auto* dep_var = new DummyVarHandle(graph->CreateControlDepVar());
+    graph->Get<GraphDepVars>(kGraphDepVars).emplace(dep_var);
+    all_reduce_op_handles[i - 1]->AddOutput(dep_var);
+    all_reduce_op_handles[i]->AddInput(dep_var);
+  }
 
-void AllReduceDepsPass::ApplyImpl(ir::Graph* graph) const {
-  auto graph_ops = ir::FilterByNodeWrapper<OpHandleBase>(*graph);
+  if (VLOG_IS_ON(10)) {
```
Comment (gongweibao) [inline paddle/fluid/framework/details/all_reduce_deps_pass.cc]: Maybe we can split it to a function.

File: paddle/fluid/framework/details/all_reduce_op_handle.cc [hunk 4]
```diff
@@ -28,7 +28,7 @@
...
```
Comment (gongweibao) [inline paddle/fluid/framework/details/all_reduce_op_handle.cc:31]: It's not useful now:
With sync:
image

Withou sync:
image

File: paddle/fluid/framework/details/build_strategy.cc [hunk 5]
```diff
@@ -132,6 +132,12 @@ class ParallelExecutorPassBuilder : public ir::PassBuilder {
       AppendPass("fuse_all_reduce_op_pass");
     }
 
+    // experimental shows that the program will be faster if append
+    // all_reduce_deps_pass here.
+    if (strategy.reduce_ == BuildStrategy::ReduceStrategy::kAllReduce) {
```
Comment (gongweibao) [inline paddle/fluid/framework/details/build_strategy.cc]: 赞。终于统一了。

File: paddle/fluid/framework/details/all_reduce_deps_pass.cc [hunk 6]
```diff
@@ -13,125 +13,186 @@
...
+      const ir::Graph& graph,
+      const std::vector<AllReduceOpHandle*>& all_reduce_op_handles) const {
+    // get vars order
+    std::map<int, std::vector<std::string>> vars =
+        GetSoredGradientsFromStaleProgram(graph);
+    std::stringstream out;
+    size_t grads_of_stale_program = 0;
+    out << "Get Order From kStaleProgramOpDescs: ";
+    for (auto& var : vars) {
+      out << "Oder " << var.first << " [";
```
Comment (gongweibao) [inline paddle/fluid/framework/details/all_reduce_deps_pass.cc]: 输出这行和141行，不利于对比。

File: paddle/fluid/framework/details/all_reduce_deps_pass.cc [hunk 7]
```diff
@@ -13,125 +13,186 @@
...
+    // Sort the current_all_reduce_op_handles according to the name of input.
+    sort(current_all_reduce_op_handles.begin(),
+         current_all_reduce_op_handles.end(),
+         [](const AllReduceOpHandle* left,
+            const AllReduceOpHandle* right) -> bool {
+           auto left_in_vars = DynamicCast<VarHandle>(left->Inputs());
+           auto right_in_vars = DynamicCast<VarHandle>(right->Inputs());
+           PADDLE_ENFORCE_GT(left_in_vars.size(), 0);
+           PADDLE_ENFORCE_EQ(left_in_vars.size(), right_in_vars.size());
+           return left_in_vars[0]->Name() < right_in_vars[0]->Name();
```
Comment (gongweibao) [inline paddle/fluid/framework/details/all_reduce_deps_pass.cc]: \>

**Review by gongweibao — CHANGES_REQUESTED** [review]

**Review by gongweibao — DISMISSED** [review]
LGTM

File: paddle/fluid/framework/details/all_reduce_op_handle.cc (hunk 2 above)
Comment (chengduoZH) [reply paddle/fluid/framework/details/all_reduce_op_handle.cc:31]: About 10%.

**Review by gongweibao — APPROVED** [review]

**Pull request merged**